lib/Nodes/CodeNode.php

```diff
@@ -4,6 +4,27 @@
 
 namespace Doctrine\RST\Nodes;
 
+/**
+ * Represents a "code node", which encompasses more than "code blocks".
+ *
+ * A code node is any block that's introduced with a :: followed
+ * by a number of lines that are considered the "value" of that node.
+ *
+ * For example, a code-block would be parsed as a CodeNode:
+ *
+ *      .. code-block:: php
+ *
+ *          // I am the first line of the value
+ *          // I am the second line
+ *
+ * But a toctree is *also* considered a CodeNode
+ *
+ *      .. toctree::
+ *          :maxdepth: 1
+ *
+ *          file
+ *          file2
+ */
```

Comment:
This was a big "ah ha!" moment for me. I had long thought it was a bug that `CodeNode` was used for some many things, like `toctree`. But apparently not - this node is just, perhaps, unfortunately named.

Reply:
:bulb: I see… what's the relationship between a CodeNode and a CodeBlock directive?

Reply:
GREAT question. 

The flow goes like this (it may be an over-simplification):

1) Document parser is actually parsing things line by line - https://github.com/doctrine/rst-parser/blob/0.3.x/lib/Parser/DocumentParser.php#L220. It first does some setup where if we're not currently already parsing something (e.g. like a table or code block) it tries to determine what this next thing will be. This part determines, for example, if the next thing might be a directive - https://github.com/doctrine/rst-parser/blob/0.3.x/lib/Parser/DocumentParser.php#L223-L289 - this step stores what the state is (table, list, directive) and, when applicable, information about what the directive is.

2) Once we are "in a block", that same `parseLine()` method (called now on the 2nd line of whatever we're in) looks at the STATE and creates different Node objects, like `CodeNode`, `ListNode`, `TableNode` etc. https://github.com/doctrine/rst-parser/blob/0.3.x/lib/Parser/DocumentParser.php#L470-L513

3a) After the Node is created in (2), IF we are NOT inside a directive, then that node is added to the document: https://github.com/doctrine/rst-parser/blob/0.3.x/lib/Parser/DocumentParser.php#L556

3b) After the Node is created in (2), IF we ARE inside a directive, it finds the correct `Directive` class and calls `process()` on it - https://github.com/doctrine/rst-parser/blob/0.3.x/lib/Parser/DocumentParser.php#L531-L537 - this is interesting because the job of this method is to ADD some Node to the Document. Basically, in the case of a directive, that Node is not automatically added to the Document, as is done for other nodes in step 3a. Instead, the `process()` method does it.

In step 3b - the `DirectiveProcces()` method - the directive ultimately should add a Node to the Document. But it can be ANY Node. In the case of the `CodeBlock` directive class (called specifically for `.. code-block::`, the original `CodeNode` created by `DocumentParser` is added to the Document - https://github.com/doctrine/rst-parser/blob/0.3.x/lib/Directives/CodeBlock.php#L55

But in the case of the `TocTree` directive class, it *extracts* information from the `CodeNode`, but ultimately creates its OWN `TocNode` and adds THAT - https://github.com/doctrine/rst-parser/blob/9f9887b282307c7cc8ff92a2c98aec39346696d9/lib/Directives/Toctree.php#L50-L53

I hope this helps! I'm trying to make public (here and in the code when I can) these... mysteries :). It's starting to make a lot more sense to me now.

Reply:
Btw, this is somewhat loosely based on the original parser (the directive process stuff however is completely specific for  this PHP-based parser).

The original parser is surprisingly well documented: https://sourceforge.net/p/docutils/code/HEAD/tree/trunk/docutils/docutils/parsers/rst/states.py#l43

Reply:
> GREAT question

I'm very gald I asked it, and I have another one: how do you find the energy to write such qualitative answers? I was certainly not expecting something that stellar!

> The original parser is surprisingly well documented

Wow that's indeed great quality too!

Reply:
> I'm very gald I asked it, and I have another one: how do you find the energy to write such qualitative 
> answers? I was certainly not expecting something that stellar!

Lol, thank you ❤️ . I want some history... in case someone else wonders this in the future!

FYI - I have reworked this comment - I now believe that when CodeNode is passed to a non-code directive, it's basically an abuse (and probably all directives should return false from `wantCode()` and we should remove/deprecate that eventually).

Reply:
After some further discovery (you can see it on the updated `CodeNode` class phpdoc, I discovered that my assumption was wrong. `CodeNode` "should" === "code block". Currently, it is ALSO passed to ANY directive that returns `true` for `wantCode()`... but my thinking is that everyone should return `false` from `wantCode()` and that method should be removed. At that point, `CodeNode` would correctly become the node for code blocks only.